src/evict/evict_lru.c

```diff
@@ -1987,6 +1995,233 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
+  WT_PAGE *last_parent, WT_EVICT_ENTRY *evict, bool *urgent_queuedp, bool *queuedp)
+{
+    WT_BTREE *btree;
+    WT_CACHE *cache;
+    WT_CONNECTION_IMPL *conn;
+    WT_PAGE *page;
+    bool modified, want_page;
+
+    btree = S2BT(session);
+    conn = S2C(session);
+    cache = conn->cache;
+    page = ref->page;
+    modified = __wt_page_is_modified(page);
+    *queuedp = false;
+
+    /* Don't queue dirty pages in trees during checkpoints. */
+    if (modified && WT_BTREE_SYNCING(btree)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_server_skip_dirty_pages_during_checkpoint);
+        return;
+    }
+
+    /*
+     * It's possible (but unlikely) to visit a page without a read generation, if we race with the
+     * read instantiating the page. Set the page's read generation here to ensure a bug doesn't
+     * somehow leave a page without a read generation.
+     */
+    if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_NOTSET)
+        __wt_cache_read_gen_new(session, page);
+
+    /* Pages being forcibly evicted go on the urgent queue. */
+    if (modified &&
+      (__wt_atomic_load64(&page->read_gen) == WT_READGEN_OLDEST ||
+        __wt_atomic_loadsize(&page->memory_footprint) >= btree->splitmempage)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_pages_queued_oldest);
+        if (__wt_page_evict_urgent(session, ref))
+            *urgent_queuedp = true;
+        return;
+    }
+
+    /*
+     * If history store dirty content is dominating the cache, we want to prioritize evicting
+     * history store pages over other btree pages. This helps in keeping cache contents below the
+     * configured cache size during checkpoints where reconciling non-HS pages can generate
+     * significant amount of HS dirty content very quickly.
+     */
+    if (WT_IS_HS(btree->dhandle) && __wt_cache_hs_dirty(session)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_pages_queued_urgent_hs_dirty);
+        if (__wt_page_evict_urgent(session, ref))
+            *urgent_queuedp = true;
+        return;
+    }
+
+    /* Pages that are empty or from dead trees are fast-tracked. */
+    if (__wt_page_is_empty(page) || F_ISSET(session->dhandle, WT_DHANDLE_DEAD))
+        goto fast;
+
+    /*
+     * Do not evict a clean metadata page that contains historical data needed to satisfy a reader.
+     * Since there is no history store for metadata, we won't be able to serve an older reader if we
+     * evict this page.
+     */
+    if (WT_IS_METADATA(session->dhandle) && F_ISSET(cache, WT_CACHE_EVICT_CLEAN_HARD) &&
+      F_ISSET(ref, WT_REF_FLAG_LEAF) && !modified && page->modify != NULL &&
+      !__wt_txn_visible_all(session, page->modify->rec_max_txn, page->modify->rec_max_timestamp)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_server_skip_metatdata_with_history);
+        return;
+    }
+
+    /* Skip pages we don't want. */
+    want_page = (F_ISSET(cache, WT_CACHE_EVICT_CLEAN) && !modified) ||
```

Comment:
Can this be placed earlier in the fuction to avoid work; or do the prior conditions evaluated take precedence?

Reply:
I think it can. It's a good idea.